Add `InlineSpan.updateAttributes`

LongCatIsLooong opened the pull request:
### Use cases in the framework:

- Service integrations in the framework such as https://github.com/flutter/flutter/blob/ad268e2f89fcfdbb58d15172b9c38a2aa4ce2acd/packages/flutter/lib/src/widgets/spell_check.dart#L312-L414, and https://github.com/flutter/flutter/issues/150068. Since `TextEditingController.buildTextSpan` returns a `TextSpan`, it's difficult to update the returned span to add new styles without a `update` method. 
- maybe: Linkify 

### Use cases outside the framework

- Updating an existing `InlineSpan` with new styles, especially when you only have the `InlineSpan` representation.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

LongCatIsLooong committed:
wip

LongCatIsLooong committed:
add tests

LongCatIsLooong committed:
comments

LongCatIsLooong committed:
comments

LongCatIsLooong committed:
typo

LongCatIsLooong committed:
export useResult from foundation

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into add-inlineSpan-updateAttributes

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +431,251 @@ abstract class InlineSpan extends DiagnosticableTree {
+  /// A convenience method, similar to [TextStyle.merge], that merges the
+  /// `textStyle` argument with the [TextStyle] related attributes in this
+  /// [InlineSpanAttributes] object.
+  TextStyle? updateTextStyle(TextStyle? textStyle) {
```
Same comment as above, update seems to indicate an in-place operation. Also, should this have the `useResult` annotation?

goderbauer commented on packages/flutter/lib/src/painting/text_span.dart:
```diff
@@ -440,6 +442,132 @@ class TextSpan extends InlineSpan implements HitTestTarget, MouseTrackerAnnotati
+      ? _slice(this, TextRange(start: clipEnd, end: textLength), newChildren)
+      : null;
+    final List<TextSpan> subtree = child2 == null
+     ? List<TextSpan>.filled(1, child1)
+     : (List<TextSpan>.filled(2, child1)..[1] = child2);
```
nit: indentation (increase by 1 space)

goderbauer commented on packages/flutter/test/painting/text_span_test.dart:
```diff
@@ -476,4 +477,156 @@ void main() {
+      addTearDown(recognizer.dispose);
+      final int length = simpleSpan.toPlainText().length;
+      final TextSpan withRecognizer = simpleSpan.updateAttributes(
+        InlineSpanAttributes(recognizer: Left(recognizer)),
```
As stated elsewhere, the `Left` in there feels out of place. When reading this code, it is unclear what this means and if you click through on `Left` it gives you a very generic description about Left that isn't helpful. In fact, I don't even know what docs to point people to if they want to understand what this Left instantiation here does.

goderbauer commented on packages/flutter/test/painting/text_span_test.dart:
```diff
@@ -476,4 +477,156 @@ void main() {
+        expect(span.recognizer, 1 <= i && i < 9 ? recognizer : null, reason: 'index = $i');
+      }
+
+      final TextSpan partiallyDeleted = withRecognizer.updateAttributes(
+        const InlineSpanAttributes(recognizer: InlineSpanAttributes.remove),
+        const TextRange(start: 5, end: 6),
+      );
```
The remove API is kinda awkward. Could we split up updateAttributes into addAttribute and removeAttribute to make this more straight-forward?

LongCatIsLooong replied:
The downside of that is we'll need two separate methods and two separate attribute sets (one for update and one for removal) for basically the same thing on `InlineSpan`. Maybe introducing sentinel values for removal is the way?

LongCatIsLooong committed:
review

LongCatIsLooong committed:
Use sentinel value

LongCatIsLooong commented:
Removed the `Either` type and switched to using a sentinel value (I wish `Either` was part of `dart:core` so we can have syntax sugar that allows us to avoid the explicit use of the `Left` constructor, so we can fix the "`copyWith` can't set a property to null" problem, among other things).

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -364,6 +371,26 @@ abstract class InlineSpan extends DiagnosticableTree {
+  /// the span of this [InlineSpan]. The implementation is responsible for
+  /// advancing the offset by this [InlineSpan]'s length in UTF16 code units,
+  /// such that the `offset` points to the end (enclusive) of this [InlineSpan]
+  /// when this method returns.
```
Maybe leave a note that you'd typically override this method, but call updateAttributes?

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -364,6 +371,26 @@ abstract class InlineSpan extends DiagnosticableTree {
+  InlineSpan updateAttributesAtOffset(covariant InlineSpanAttributes newAttributes, TextRange textRange, Accumulator offset);
+
+  /// Creates a new [InlineSpan] from this [InlineSpan] by applying
+  /// `newAttributes` to the given [TextRange] within this span.
```
Same here, maybe leave a comment that subclasses should override updateAttributesAtOffset instead of this one?

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +434,287 @@ abstract class InlineSpan extends DiagnosticableTree {
+///
+/// See also:
+///
+///  * [InlineSpan.updateAttributesAtOffset], which takes an [InlineSpanAttributes], and
```
Here and elsewhere in the docs: Should the docs link to `updateAttributes` instead?

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +434,287 @@ abstract class InlineSpan extends DiagnosticableTree {
+///
+/// All attributes are nullable, and default to `null`. Attributes set to `null`
+/// in an [InlineSpanAttributes] will not be updated.
+///
```
If that situation is uncommon (you probably have a better feeling than I for that) we should still document what's possible and what isn't.

LongCatIsLooong replied:
> remove the fontFamily

Ah that sounds like a legit use case. There's no way to use the "default" font family unless it's set to null. 

> it's gonna be inferred from e.g. a sounding DefaultTextStyle 

It happens in the `Text` widget I think. But yeah to support removing `fontSize` (such that the text uses the ambient DefaultTextStyle fontSize) we would need a different sentinel value.

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +434,287 @@ abstract class InlineSpan extends DiagnosticableTree {
+  /// replaces existing recognizers (if any) within the given range with the
+  /// specified recognizer.
+  ///
+  /// When this value is set to [remove], the [InlineSpan.updateAttributesAtOffset]
```
Wondering: Could we use null to indicate "remove" by using `remove` (under another name) as the default value for this property? That default value could maybe even have a private type so we don't need to export RemoveInlineSpanAttribute.

LongCatIsLooong replied:
Since the sentinel values aren't exposed, whether an attribute should be updated would become a compile-time decision? 
E.g.:
```dart
final Color? maybeColor = ...;
if (maybeColor == null) {
  updateAttributes(.... other attributes);
} else { 
  updateAttributes(.... other attributes, color: maybeColor);
}
```
It will quickly become unscaleable configuration matrix I think,  not great for APIs that want to wrap this API.

goderbauer commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +434,287 @@ abstract class InlineSpan extends DiagnosticableTree {
+  /// [TextSpan.onEnter] can not be set from a non-null value to null using this
+  /// attribute. Considering setting [TextSpan.onEnter] to a function that does
+  /// nothing if you want to unset the callback from a [TextSpan].
+  final PointerEnterEventListener? onEnter;
```
That default value could also work for these, reducing the number of remove options from 3 (unremovable, removable with `remove`, removable by setting to a no-op) to 2 (unremovable, use null to indicate remove).

LongCatIsLooong commented on packages/flutter/lib/painting.dart:
```diff
@@ -50,7 +50,7 @@ export 'src/painting/image_decoder.dart';
 export 'src/painting/image_resolution.dart';
 export 'src/painting/image_stream.dart';
-export 'src/painting/inline_span.dart';
+export 'src/painting/inline_span.dart' hide RemoveInlineSpanAttribute;
```
I would still prefer hiding the type here since its only inhabitant is a sentinel value that only supports the `identical` operation.

LongCatIsLooong committed:
fix wrong refs

chunhtai commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +431,251 @@ abstract class InlineSpan extends DiagnosticableTree {
+      && lineThrough == null
+      && decorationColor == null
+      && decorationStyle == null
+      && decorationThickness == null;
```
+1

LongCatIsLooong replied:
Hmm the reason I introduced this class (instead of doing something with TextStyle`) is that `TextStyle` has its own copy with problems. But maybe we should fix the problem on TextStyle instead so this class doesn't have to maintain as many fields.

chunhtai commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -407,3 +439,287 @@ abstract class InlineSpan extends DiagnosticableTree {
+  /// [InlineSpan.updateAttributes] does not mutate the target
+  /// [InlineSpan]. Rather, when [recognizer] is set to [remove], that method
+  /// returns a new [InlineSpan] with [recognizer] set to null.
+  static const RemoveInlineSpanAttribute remove = _PoorMansBottomType._();
```
looks like it is also used in text_span.dart, but I think it is still better to ask people to use `remove` property to clean up the field.

Also have we consider merging these span libraries into one file? it feels unecessary to separate them and expose a bunch of internal method...

LongCatIsLooong replied:
If I understand correctly, `_somePrivateDefaultValue` is a sentinel value that means "don't change the attribute" right? Then we'll need to use `null` for "set the value back to null`.

Since historically we use `null` for this, I feel we should stick to `null` for "keeping the attribute as-is"?

chunhtai reviewed (commented):
the updating function mostly looks good to me

LongCatIsLooong commented:
Still evaluating some alternatives as I would like to expose more lower level text layout apis, and at the same time some hacks in the PR made the API less appealing (e.g., the removal problem).

gnprice commented on packages/flutter/lib/src/painting/inline_span.dart:
```diff
@@ -364,6 +379,15 @@ abstract class InlineSpan extends DiagnosticableTree {
+  /// from the start of the root span to sub-spans , it should be set to null
+  /// (the default) if you are not implementing an `InlineSpan` subclass.
+  @useResult
+  InlineSpan updateAttributes(covariant InlineSpanAttributes newAttributes, TextRange textRange, { Accumulator? offset });
```
DBC: To my eye "apply" can equally well describe something that mutates, or something that returns its result instead. After all it's a generic term used for function calls, and functions can be either mutating or non-mutating.

Similarly I read "merge" as describing how the result relates to the inputs, but leaving open whether something gets mutated into the result state or instead everything's immutable and the result gets returned.

But "update" on the other hand specifically sounds to me like it's talking about changing something, mutating it.

So I'd read "applyAttributes" or "mergeAttributes" as leaving unstated whether the span gets mutated, which is OK because either the method's doc or a good understanding of InlineSpan in general resolves the uncertainty; but "updateAttributes" as misleading because it seems to specifically say something that isn't actually what's intended.

Piinks commented:
(Triage) Confirmed with @LongCatIsLooong this is still on their radar

Piinks commented:
Hello from triage again! Is this still something you'd like to work on?

Piinks commented:
Greetings from stale PR triage (again)! 👋 
Should we close this PR?
